### PR TITLE
[SPARK-35611][SS] Introduce the strategy on mismatched offset for start offset timestamp on Kafka data source

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to introduce the strategy on mismatched offset for start offset timestamp on Kafka data source.

Please read the section `Why are the changes needed?` to understand the rationalization of the functionality.

This would be pretty much helpful for the case where there's a skew between partitions and some partitions have older records.

* AS-IS: Spark simply fails the query and end users have to deal with workarounds requiring manual steps.
* TO-BE: Spark will assign the latest offset for these partitions, so that Spark can read newer records from these partitions in further micro-batches.

To retain the existing behavior and also give some help for the proposed "TO-BE" behavior, we'd like to introduce the strategy on mismatched offset for start offset timestamp to let end users choose from them.

The strategy will be added as source option, to ensure end users set the behavior explicitly (otherwise simply "known" default value).

* New source option to be added: startingOffsetsByTimestampStrategy
* Available values: `error` (fail the query as referred as AS-IS), `latest` (set the offset to the latest as referred as TO-BE)

Doc changes are following:

ES-106042-doc-screenshot-1
ES-106042-doc-screenshot-2

### Why are the changes needed?

We encountered a real-world case Spark fails the query if some of the partitions don't have matching offset by timestamp.

This is intended behavior to avoid bring unintended output for some cases like:

* timestamp 2 is presented as timestamp-offset, but the some of partitions don't have the record yet
* record with timestamp 1 comes "later" in the following micro-batch

which is possible since Kafka allows to specify the timestamp in record.

Here the unintended output we talked about was the risk of reading record with timestamp 1 in the next micro-batch despite the option specifying timestamp 2.

But for many cases end users just suppose timestamp is increasing monotonically with wall clocks are all in sync, and current behavior blocks these cases to make progress.

### Does this PR introduce _any_ user-facing change?

Yes, but not a breaking change. It's up to end users to choose the behavior which the default value is "error" (current behavior). And it's a source option (not config) so they need to explicitly set the behavior to let the functionality takes effect.

### How was this patch tested?

New UTs.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @gaborgsomogyi @viirya @xuanyuanking Please take a look, thanks!

**Comment by viirya** [comment]
Let me know if I understand correctly or not. So it sounds like when start offset timestamp cannot be found on Kafka, Spark will turn to read latest offset instead. I think it makes sense there is an option for end users to avoid query failure in that case. Just wondering, is latest offset the best option? If the start offset timestamp is far before latest offset, does it still make sense to retrieve latest offset?

For example, partition 1 is able to get record with timestamp 1, but partition 2 returns unmatched offset. With this option, we retrieve latest offset from partition 2 instead. But the latest offset could be with timestamp 1000, though there are some records after timestamp 1 before timestamp 1000?

**Comment by HeartSaVioR** [comment]
> Let me know if I understand correctly or not. So it sounds like when start offset timestamp cannot be found on Kafka, Spark will turn to read latest offset instead. I think it makes sense there is an option for end users to avoid query failure in that case. Just wondering, is latest offset the best option? If the start offset timestamp is far before latest offset, does it still make sense to retrieve latest offset?
> 
> For example, partition 1 is able to get record with timestamp 1, but partition 2 returns unmatched offset. With this option, we retrieve latest offset from partition 2 instead. But the latest offset could be with timestamp 1000, though there are some records after timestamp 1 before timestamp 1000?

Yeah that is quite tricky to understand - the fact is, the functionality of "offset by timestamp" never be able to guarantee that there should be NO records in further batches less than specified timestamp. IIUC there's no ordering requirement in Kafka - even the default behavior of Kafka doesn't guarantee such ordering; by default, the timestamp is marked from producer side, which would suffer with the known clock issue. Even you can manually set the timestamp in record, which means that the timestamp can be the semantic of "event time", and now you can't predict about the ordering at all.

That's why we had to explain in doc that we are simply calling Kafka's offset by timestamp and leveraging the result. Now you'll realize the fact that the problem exists even if the matched offset is found. The functionality of "offset by timestamp" should be used with understanding of this.

So there's no best option works for every cases, as even the normal case can be broken if you consider the semantic of "offset by timestamp" as no records will be read which have less than the specified timestamp. The data itself should respect the ordering, or end users should tolerate some missing records.

**Comment by HeartSaVioR** [comment]
Ah also worth to say, IIUC, "latest" does not mean "current latest record in partition". The semantic of "latest" is actually `(current latest record in partition + 1)` so the further input records will be read. The semantic already works for other offset options - you probably read nothing if you specify "latest" as starting offset in first batch.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 1]
```diff
@@ -1456,8 +1456,7 @@ abstract class KafkaSourceSuiteBase extends KafkaSourceTest {
       testFromSpecificOffsets(
         topic,
         failOnDataLoss = failOnDataLoss,
-        "assign" -> assignString(topic, 0 to 4),
-        "failOnDataLoss" -> failOnDataLoss.toString)
```
Comment (xuanyuanking) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:1460]: IIUC, this is unrelated to this feature, just delete the duplicated setting for `failOnDataLoss`?

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 2]
```diff
@@ -1568,6 +1565,137 @@ abstract class KafkaSourceSuiteBase extends KafkaSourceTest {
...
+          0, KafkaSourceOffset(partitions.toMap), streamingTimeout.toMillis)
+      },
+      CheckNewAnswer(23, 24, 25)
+    )
+  }
+
+  private def assertQueryFailOnStartOffsetStrategyAsError(df: Dataset[_]): Unit = {
+    // In continuous mode, the origin exception is not caught here unfortunately, so we have to
+    // stick with checking general exception instead of verifying IllegalArgumentException.
+    intercept[Exception] {
```
Comment (xuanyuanking) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:1694]: If we have to stick with checking general exception, let's add more check on the error messages.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderConsumer.scala [hunk 3]
```diff
@@ -270,6 +262,35 @@ private[kafka010] class KafkaOffsetReaderConsumer(
...
+      isStartingOffsets: Boolean,
+      strategyOnNoMatchStartingOffset: StrategyOnNoMatchStartingOffset.Value,
+      partitionTimestampFn: TopicPartition => Long): Map[TopicPartition, Long] = {
+
+    tpToOffsetMap.map { case (tp, offsetSpec) =>
+      val offset = if (offsetSpec == null) {
+        if (isStartingOffsets) {
+          strategyOnNoMatchStartingOffset match {
+            case StrategyOnNoMatchStartingOffset.ERROR =>
+              throw new IllegalArgumentException("No offset " +
```
Comment (xuanyuanking) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderConsumer.scala:276]: The exception type changed. Before is `AssertionError`, now is `IllegalArgumentException`. Maybe we can keep the same type to avoid user behavior changes?

Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderConsumer.scala:276]: Ah OK we used assert, not require. Good point.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala (hunk 1 above)
Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:1460]: Yes, we set the same config unnecessarily twice.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala (hunk 2 above)
Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:1694]: In continuous mode, the error message is completely different than we expect. This is a log message I get from `exc.getMessage()`:

```
21/06/18 22:10:20.680 ScalaTest-run-running-KafkaContinuousSourceSuite WARN KafkaContinuousSourceSuite: 
IncrementalExecution was not created: The code passed to eventually never returned normally. Attempted 1794 times over 30.009420817000002 seconds. Last failure message: null equaled null.
```

The actual exception is logged, but we can't intercept the original exception here.

```
21/06/18 22:09:53.747 stream execution thread for [id = bc35e8a2-077e-4949-b8b6-47670d1c057f, runId = d528f706-e553-4603-b022-fc0fe7cc0432] ERROR ContinuousExecution: Query [id = bc35e8a2-077e-4949-b8b6-47670d1c057f, runId = d528f706-e553-4603-b022-fc0fe7cc0432] terminated with error
java.lang.AssertionError: No offset matched from request of topic-partition topic-31-4 and timestamp 1624021786438.
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.$anonfun$readTimestampOffsets$1(KafkaOffsetReaderConsumer.scala:280)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:400)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:728)
	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.readTimestampOffsets(KafkaOffsetReaderConsumer.scala:271)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.$anonfun$fetchSpecificTimestampBasedOffsets$4(KafkaOffsetReaderConsumer.scala:227)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.$anonfun$fetchSpecificOffsets0$1(KafkaOffsetReaderConsumer.scala:305)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.$anonfun$partitionsAssignedToConsumer$2(KafkaOffsetReaderConsumer.scala:561)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.$anonfun$withRetriesWithoutInterrupt$1(KafkaOffsetReaderConsumer.scala:594)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.util.UninterruptibleThread.runUninterruptibly(UninterruptibleThread.scala:77)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.withRetriesWithoutInterrupt(KafkaOffsetReaderConsumer.scala:593)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.$anonfun$partitionsAssignedToConsumer$1(KafkaOffsetReaderConsumer.scala:547)
	at org.apache.spark.util.UninterruptibleThreadRunner.runUninterruptibly(UninterruptibleThreadRunner.scala:48)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.partitionsAssignedToConsumer(KafkaOffsetReaderConsumer.scala:547)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.fetchSpecificOffsets0(KafkaOffsetReaderConsumer.scala:301)
	at org.apache.spark.sql.kafka010.KafkaOffsetReaderConsumer.fetchSpecificTimestampBasedOffsets(KafkaOffsetReaderConsumer.scala:233)
	at org.apache.spark.sql.kafka010.KafkaContinuousStream.initialOffset(KafkaContinuousStream.scala:74)
	at org.apache.spark.sql.execution.streaming.continuous.ContinuousExecution$$anonfun$3.$anonfun$applyOrElse$4(ContinuousExecution.scala:170)
...
```

Comment (xuanyuanking) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:1694]: Ah, got it.

**Review by xuanyuanking — APPROVED** [review]
LGTM

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala [hunk 4]
```diff
@@ -66,30 +67,40 @@ private[kafka010] trait KafkaOffsetReader {
...
+   * - isStartingOffsets = true  => implementation should follow the strategy on non-matching
+   *                                starting offset, passed as `strategyOnNoMatchStartingOffset`
    *
    * @param partitionTimestamps the timestamp per topic-partition.
-   * @param failsOnNoMatchingOffset whether to fail the query when no matched offset can be found.
    */
   def fetchSpecificTimestampBasedOffsets(
       partitionTimestamps: Map[TopicPartition, Long],
-      failsOnNoMatchingOffset: Boolean): KafkaSourceOffset
+      isStartingOffsets: Boolean,
```
Comment (viirya) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala:82]: `isStartingOffsets` here actually means the starting offset of the query, right? We have another concept that is the start offset for current batch (e.g. `isStartingOffsets` in `fetchPartitionOffsets`), do you think it is better to rename this as `isQueryStartingOffsets`?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderAdmin.scala [hunk 5]
```diff
@@ -135,12 +136,12 @@ private[kafka010] class KafkaOffsetReaderAdmin(
...
-          failsOnNoMatchingOffset = isStartingOffsets).partitionToOffsets
-      case GlobalTimestampRangeLimit(timestamp) =>
-        fetchGlobalTimestampBasedOffsets(timestamp,
-          failsOnNoMatchingOffset = isStartingOffsets).partitionToOffsets
+      case SpecificTimestampRangeLimit(partitionTimestamps, strategyOnNoMatchingStartingOffset) =>
+        fetchSpecificTimestampBasedOffsets(partitionTimestamps, isStartingOffsets,
+          strategyOnNoMatchingStartingOffset).partitionToOffsets
+      case GlobalTimestampRangeLimit(timestamp, strategyOnNoMatchingStartingOffset) =>
+        fetchGlobalTimestampBasedOffsets(timestamp, isStartingOffsets,
+          strategyOnNoMatchingStartingOffset).partitionToOffsets
```
Comment (viirya) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderAdmin.scala:144]: Hmm, but the `isStartingOffsets` here is actually the starting offset of batch?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderAdmin.scala [hunk 6]
```diff
@@ -208,25 +205,47 @@ private[kafka010] class KafkaOffsetReaderAdmin(
...
+        if (isStartingOffsets) {
+          strategyOnNoMatchStartingOffset match {
+            case StrategyOnNoMatchStartingOffset.ERROR =>
+              throw new IllegalArgumentException("No offset " +
+                s"matched from request of topic-partition $tp and timestamp " +
+                s"${partitionTimestampFn(tp)}.")
+
+            case StrategyOnNoMatchStartingOffset.LATEST =>
+              KafkaOffsetRangeLimit.LATEST
+          }
```
Comment (viirya) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderAdmin.scala:237]: Looks like this checks the strategy for each start offsets per batch? I'm a bit confused. Isn't the `startingOffsetsByTimestamp` only for `The start point of timestamp when a query is started ...`?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderAdmin.scala (hunk 5 above)
Comment (viirya) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderAdmin.scala:144]: Oh, I see. `fetchPartitionOffsets` is only called for batch query.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderAdmin.scala (hunk 6 above)
Comment (viirya) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReaderAdmin.scala:237]: This is the same. nvm. Although `fetchSpecificTimestampBasedOffsets` is called in batch, and streaming batch queries, for streaming batch it is only called for initial partition offsets.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks @HeartSaVioR for the PR and @xuanyuanking for the review! Merging to master.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!
